audit_log/admin.py

```diff
@@ -0,0 +1,228 @@
+import json
+import logging
+
+from django.contrib import admin
+from django.core.paginator import Paginator
+from django.db import connection
+from django.utils.functional import cached_property
+from django.utils.html import escape
+from django.utils.safestring import mark_safe
+from django.utils.translation import gettext_lazy as _
+
+from audit_log.enums import Operation, Role
+from audit_log.models import AuditLogEntry
+from audit_log.services import audit_log_service, create_api_commit_message_from_request
+
+logger = logging.getLogger(__name__)
+
+
+class AuditLogModelAdminMixin:
+    def get_queryset(self, request):
+        """Write audit logs of the queryset read."""
+        # NOTE: Fetching a single object's edit page reuses the list view queryset
+        return (
+            super()
+            .get_queryset(request)
+            .with_audit_log_and_request(
+                request=request,
+                operation=Operation.READ.value,
+            )
+        )
+
+    def save_model(self, request, obj, form, change):
+        """
+        Given a model instance save it to the database.
+        """
+        # TODO: add a change log to the audit log message
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.UPDATE.value if change else Operation.CREATE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().save_model(request, obj, form, change)
+
+    def delete_model(self, request, obj):
+        """
+        Given a model instance delete it from the database.
+        """
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.DELETE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().delete_model(request, obj)
+
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        super().delete_queryset(
+            request,
+            queryset.with_audit_log_and_request(
+                request=request, operation=Operation.DELETE.value
+            ),
+        )
+
+
+class AuditLogMessageOperationListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("operation")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "operation"
+
+    options = [operation.value for operation in Operation]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
```

Comment:
```suggestion
        return [(option, option.upper()) for option in self.options]
```
to get rid of the `lambda`, list comprehension reads quite nicely here IMHO.

Reply:
Updated